### PR TITLE
Add e2e regression tests for the kubelet being secure

### LINKED ISSUE #838: Add e2e regression tests for the kubelet being secure
As part of https://github.com/kubernetes/kubeadm/issues/732, https://github.com/kubernetes/kubeadm/issues/650 and https://github.com/kubernetes/kubernetes/pull/63881 we should make sure the following things in our end-to-end testing:
1. The kubelet cAdvisor port (4194) can't be reached, neither via the API server proxy nor directly on the public IP address
2. The kubelet read-only port (10255) can't be reached, neither via the API server proxy nor directly on the public IP address
3. The kubelet can delegate ServiceAccount tokens to the API server
4. The kubelet's main port (10250) has both authentication (should fail with no credentials) and authorization (should fail with insufficient permissions) set-up

These e2e tests, which I preliminarily propose to host under `[sig-cluster-lifecycle] [Feature:KubeletSecurity]`, would be run by any kubeadm e2e suite running against v1.11+ clusters.
These test are super important to make sure no kubeadm version regresses security-wise by accident.

@dixudx is working on creating these tests, thank you a lot!

FYI @kubernetes/sig-node-proposals @kubernetes/sig-auth-proposals

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:
This PR does,

1. The kubelet cAdvisor port (4194) can't be reached, neither via the API server proxy nor directly on the public IP address
2. The kubelet read-only port (10255) can't be reached, neither via the API server proxy nor directly on the public IP address
3. The kubelet can delegate ServiceAccount tokens to the API server
4. The kubelet's main port (10250) has both authentication (should fail with no credentials) and authorization (should fail with insufficient permissions) set-up

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes kubernetes/kubeadm#838

**Special notes for your reviewer**:
/cc luxas tallclair 
**Release note**:

```release-note
Add e2e regression tests for the kubelet being secure
```

### CODE REVIEW

File: test/e2e/auth/node_authn.go
Comment (luxas) [inline test/e2e/auth/node_authn.go]: I don't think we can do this quite yet. Requires @kubernetes/sig-architecture-pr-reviews approval. Let's stick with `It` now, and do the one-line change to Conformance later if the request is approved. Please do add a comment to follow-up.

Comment (luxas) [reply test/e2e/auth/node_authn.go]: suggest: `It("Should make the kubelet's main port 10250 enforce authentication for client requests", `

File: test/e2e/auth/node_authn.go
Comment (luxas) [inline test/e2e/auth/node_authn.go]: Also require the returned SA to have one or more Secrets so the call later won't panic

File: test/e2e/auth/node_authn.go
Comment (luxas) [inline test/e2e/auth/node_authn.go]: Add a comment why you do this. (I think it's to make sure the ServiceAccount admission controller is enabled etc., so secret generation on SA creation works, right?)

File: test/e2e/auth/node_authn.go
Comment (luxas) [inline test/e2e/auth/node_authn.go]: Will this actually work when you enforce RBAC? Don't you have to create RBAC rules to make the SA able to access the kubelet proxy? Also, I'd expect you to talk to the kubelet directly:

```go
_, err = c.CoreV1().RESTClient().Get().
			Context(ctx).
			Resource("nodes").
			SubResource("proxy").
			Name(fmt.Sprintf("%v:%v", nodeName, ports.KubeletPort)).
			Suffix("metrics").
			Do().Raw()
```

File: test/e2e/auth/node_authz.go
Comment (luxas) [inline test/e2e/auth/node_authz.go]: Don't add conformance here please. You can add a comment though to maybe do it in the future though.
cc @tallclair

File: test/e2e/lifecycle/kubelet_security.go
Comment (luxas) [inline test/e2e/lifecycle/kubelet_security.go]: Add a comment why you need this?

File: test/e2e/lifecycle/kubelet_security.go
Comment (luxas) [inline test/e2e/lifecycle/kubelet_security.go]: suggest: `should not be able to proxy to the readonly kubelet port 10255 using proxy subresource` to be consistent with what you have below

File: test/e2e/lifecycle/kubelet_security.go
Comment (luxas) [inline test/e2e/lifecycle/kubelet_security.go]: suggest: `should not have port %d open on its public IP address`

File: test/e2e/lifecycle/kubelet_security.go
Comment (luxas) [inline test/e2e/lifecycle/kubelet_security.go]: Can you test that the port is closed on all addresses just to be safe?

**Review by luxas — COMMENTED** [review]
This looks great overall.
Thanks for this valuable contribution @dixudx :+1:!
I left a couple of initial comments.

I'd love a review from @kubernetes/sig-auth-pr-reviews, and a comment from @kubernetes/sig-architecture-pr-reviews regarding eventually graduating `It("Should make the kubelet's main port 10250 enforce authentication for client requests")` a Conformance test as @tallclair commented in https://github.com/kubernetes/kubeadm/issues/838#issuecomment-390789959

File: test/e2e/auth/node_authn.go
Comment (dixudx) [reply test/e2e/auth/node_authn.go]: Correct. Add a comment is better.

File: test/e2e/auth/node_authn.go
Comment (dixudx) [reply test/e2e/auth/node_authn.go]: Yes. Good catch.

**Review by luxas — APPROVED** [review]
Thanks a lot @dixudx!!
I'm fine for the moment, the high level functionality is there, deferring the rest of the review to SIG Auth reviewers.
I'm gonna try running these locally in a minute to make sure everything actually works :smile:

/approve
/lgtm

File: test/e2e/auth/node_authn.go
Comment (liggitt) [inline test/e2e/auth/node_authn.go]: this isn't testing what you think it is. you're hitting the API server anonymously, and the apiserver is rejecting you. also, in most configurations, it would actually return a forbidden error, not an unauthorized error.

File: test/e2e/auth/node_authn.go
Comment (liggitt) [inline test/e2e/auth/node_authn.go]: when fetching the token object into a go struct, you don't need to base64 decode secret data... you can use the token key directly

File: test/e2e/auth/node_authn.go
Comment (liggitt) [inline test/e2e/auth/node_authn.go]: same comment applies... you're not testing the kubelet endpoints here, you're testing the API server.

File: test/e2e/auth/node_authz.go
Comment (liggitt) [inline test/e2e/auth/node_authz.go]: same comment, this is likely testing the apiserver, not the kubelet. I'm also surprised the client was rejected here... the framework client is typically highly privileged

**Comment by liggitt** [comment]
while I like the idea of exercising the described tests, these e2e tests don't actually do that. the only reason CI is green with them added is because they're behind feature tags that don't actually run them (or the API server happens to return the same response we were hoping to see from the kubelet)

**Comment by tallclair** [comment]
I recommend proxying the tests through shell commands run in a container and pointed at the kubelets cluster-internal IP address. See the apparmor test for an example of how you might do this:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/common/apparmor.go#L60

**Comment by dixudx** [comment]
ping @liggitt @luxas @tallclair for another look

**Comment by timothysc** [comment]
> while I like the idea of exercising the described tests, these e2e tests don't actually do that. the only reason CI is green with them added is because they're behind feature tags that don't actually run them (or the API server happens to return the same response we were hoping to see from the kubelet) 

We would add them to a kubeadm suite, and when mainline is happy we can eliminate the "Feature" option.

**Comment by liggitt** [comment]
>> while I like the idea of exercising the described tests, these e2e tests don't actually do that. the only reason CI is green with them added is because they're behind feature tags that don't actually run them (or the API server happens to return the same response we were hoping to see from the kubelet)
>
> We would add them to a kubeadm suite, and when mainline is happy we can eliminate the "Feature" option.

That comment was in response to the previous content of the PR which was making calls to the API server, not the kubelet.

**Comment by timothysc** [comment]
/approve 

defer to sig-auth folks on lgtm

File: test/e2e/auth/node_authn.go
Comment (tallclair) [inline test/e2e/auth/node_authn.go]: How about instead of doing this manually, just allow automount and read it from the volume:
```
curl ... -header "Authorization: Bearer `cat /var/run/secrets/kubernetes.io/serviceaccount/token`"
```

File: test/e2e/lifecycle/kubelet_security.go
Comment (tallclair) [inline test/e2e/lifecycle/kubelet_security.go]: nit: body is never used; don't return it.

File: test/e2e/lifecycle/kubelet_security.go
Comment (tallclair) [inline test/e2e/lifecycle/kubelet_security.go]: nit: use ports.KubeletReadOnlyPort instead of 10255

File: test/e2e/lifecycle/kubelet_security.go
Comment (tallclair) [inline test/e2e/lifecycle/kubelet_security.go]: See [NodeProxyRequest](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L4497) for a better approach (or maybe refactor that to have a version that allows you to specify the port)

File: test/e2e/lifecycle/kubelet_security.go
Comment (tallclair) [inline test/e2e/lifecycle/kubelet_security.go]: reuse the node address accumulation from the other test file

File: test/e2e/lifecycle/kubelet_security.go
Comment (tallclair) [inline test/e2e/lifecycle/kubelet_security.go]: Use `DialTimeout`

File: test/e2e/lifecycle/kubelet_security.go
Comment (tallclair) [inline test/e2e/lifecycle/kubelet_security.go]: prefer: `framework.Failf("port %d is not disabled", port)`

File: test/e2e/auth/node_authn.go
Comment (dixudx) [reply test/e2e/auth/node_authn.go]: Yeah, this is better. Good advice.

Comment (tallclair) [reply test/e2e/auth/node_authn.go]: I don't understand this.... what do you mean "will get rendered"?

File: test/e2e/auth/node_authn.go
Comment (tallclair) [inline test/e2e/auth/node_authn.go]: `CreateSync` handles this for you. You can remove this and the next 2 lines.

File: test/e2e/auth/node_authn.go
Comment (tallclair) [inline test/e2e/auth/node_authn.go]: nit: remove this, it's unused.

File: test/e2e/auth/node_authn.go
Comment (tallclair) [inline test/e2e/auth/node_authn.go]: Remove. The framework pod client handles it.

File: test/e2e/auth/node_authn.go
Comment (tallclair) [inline test/e2e/auth/node_authn.go]: I get `Unauthorized`. I think a more robust approach would be to just look for a 401 or 403 status code:

```
curl -s -o /dev/null -w "%{http_code}" ... | grep "^40[13]$"
```

Then you can just check the exit status of exec command too.

File: test/e2e/lifecycle/kubelet_security.go
Comment (tallclair) [inline test/e2e/lifecycle/kubelet_security.go]: Internal IPs are meaningless from the e2e test. If you want to test internal IPs it must be done from a pod (or something on the same private network).

**Comment by jberkus** [comment]
Are these tests critical to have for 1.11?  This PR doesn't seem ready to merge.

**Comment by tallclair** [comment]
Yeah, this isn't a conformance test, so I don't think it needs to be tied to a version. We can merge it when it's ready.

**Comment by dixudx** [comment]
@jberkus @tallclair Addressed the comments. PTAL. Thanks.

File: test/e2e/auth/node_authn.go
Comment (tallclair) [inline test/e2e/auth/node_authn.go]: This comment only applies to the test runner. Since you're execing both the tests through a pod, you can test the internal IPs here.

File: test/e2e/auth/node_authn.go
Comment (tallclair) [inline test/e2e/auth/node_authn.go]: nit: s/fail/reject requests/

File: test/e2e/auth/node_authn.go
Comment (tallclair) [inline test/e2e/auth/node_authn.go]: I'd prefer to check for 401 OR 403, that way you're not depending on a specific implementation as much.

File: test/e2e/auth/node_authn.go
Comment (tallclair) [inline test/e2e/auth/node_authn.go]: curl doesn't know how to interpret backticks - you need to run this command in a shell. I.e.:
```
[]string{"sh", "-c", "curl -sI --insecure --header "Authorization: Bearer `cat /var/run/secrets/kubernetes.io/serviceaccount/token`" + url}
```

File: test/e2e/auth/node_authn.go
Comment (tallclair) [inline test/e2e/auth/node_authn.go]: ... or 401

File: test/e2e/lifecycle/kubelet_security.go
Comment (tallclair) [inline test/e2e/lifecycle/kubelet_security.go]: Are you sure the path should include `proxy` here?

**Comment by tallclair** [comment]
There seems to be some confusion about what internal IPs and external IPs are.

External IPs are available to the entire internet.
Internal IPs are from a [private address space](https://en.wikipedia.org/wiki/Private_network#Private_IPv4_address_spaces), meaning they are only meaningful within the private network. In our case, that means running in the cluster.

The E2E test runner runs outside the cluster being tested, which means that it can only see the external IP addresses (_actually, it can see the internal IP addresses, but they are resolved within the test runner's private network, so they probably resolve to different devices, if at all_). When you run a pod, it runs _in_ the cluster, meaning the pod _can_ see the internal addresses, this is why an tests that look at the internal IPs should be run through a pod.

Hopefully that clears it up?

File: test/e2e/auth/node_authn.go
Comment (tallclair) [inline test/e2e/auth/node_authn.go]: Same as above comment.

File: test/e2e/auth/node_authn.go
Comment (tallclair) [inline test/e2e/auth/node_authn.go]: `Expect(result).To(And(BeNumerically(">=", 200), BeNumerically("<", 300)), ...)`

Comment (tallclair) [reply test/e2e/auth/node_authn.go]: I don't particularly care for the gomega DSL, so feel free to do these checks explicitly, but semantically this is the condition you should check for.

File: test/e2e/auth/node_authn.go
Comment (tallclair) [inline test/e2e/auth/node_authn.go]: This comment still doesn't make sense to me. Why did you drop the external IPs?

File: test/e2e/lifecycle/kubelet_security.go
Comment (tallclair) [inline test/e2e/lifecycle/kubelet_security.go]: The 0th node is usually the master, which might be configured differently. Still a good idea to check it though, but maybe pick a few (or all of them)?

File: test/e2e/auth/node_authn.go
Comment (tallclair) [reply test/e2e/auth/node_authn.go]: You might be able to simplify this to just: `Expect(result).To(Equal("200"), ...)`

Comment (dixudx) [reply test/e2e/auth/node_authn.go]: Here we're using a new SA token when sending requests to the kubelet. The expected result should be 403, and authentication is okay.

> Expect(result).To(Equal("200"), ...)

@tallclair I don't quite understand where `200` comes.

File: test/e2e/auth/node_authn.go
Comment (dixudx) [reply test/e2e/auth/node_authn.go]: So we add external IPs back again? I was testing both internal IPs and external IPs before.

File: test/e2e/lifecycle/kubelet_security.go
Comment (dixudx) [reply test/e2e/lifecycle/kubelet_security.go]: Actually all the nodes here are schedulable and ready. Usually the master node is tainted, which will be filtered due to unschedulable.

So the 0th node is usually not the master. And 0th node is used widely in our e2e test scenarios.

https://github.com/kubernetes/kubernetes/blob/ad9722c453d28fb7eac4f21c82f60767df3bcb51/test/e2e/node/mount_propagation.go#L86-L89

**Comment by liggitt** [comment]
per https://github.com/kubernetes/kubernetes/pull/64140#issuecomment-394021455 this isn't tied to v1.11 or blocking the release

/milestone clear

File: test/e2e/auth/node_authn.go
Comment (tallclair) [inline test/e2e/auth/node_authn.go]: This is overwriting the ExternalIPs

Comment (dixudx) [reply test/e2e/auth/node_authn.go]: Wow, what a big mistake. My fault.

**Comment by dixudx** [comment]
@tallclair @luxas Seems the bot does not want to lgtm.

**Pull request merged**
